From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stefan Eissing <stefan.eissing@greenbytes.de>
Date: Thu, 4 Nov 2021 10:01:37 +0100
Subject: [PATCH 21/21] Patch mod_http2 to fix issue with hanging http2 threads

* A regression in v1.15.24 was fixed that could lead to httpd child processes not being terminated on a graceful reload or when reaching MaxConnectionsPerChild. When unprocessed h2 requests were queued at the time, these could stall. See #212.

 modules/http2/h2_session.c | 4 ++--
 modules/http2/h2_workers.c | 2 --
 2 files changed, 2 insertions(+), 4 deletions(-)
---
 modules/http2/h2_session.c | 4 ++--
 modules/http2/h2_workers.c | 2 --
 2 files changed, 2 insertions(+), 4 deletions(-)

diff --git a/modules/http2/h2_session.c b/modules/http2/h2_session.c
index dc883b5..a97c0cb 100644
--- a/modules/http2/h2_session.c
+++ b/modules/http2/h2_session.c
@@ -275,7 +275,7 @@ static int on_begin_headers_cb(nghttp2_session *ngh2,
                                const nghttp2_frame *frame, void *userp)
 {
     h2_session *session = (h2_session *)userp;
-    h2_stream *s;
+    h2_stream *s = NULL;
     
     /* We may see HEADERs at the start of a stream or after all DATA
      * streams to carry trailers. */
@@ -284,7 +284,7 @@ static int on_begin_headers_cb(nghttp2_session *ngh2,
     if (s) {
         /* nop */
     }
-    else {
+    else if (session->local.accepting) {
         s = h2_session_open_stream(userp, frame->hd.stream_id, 0);
     }
     return s? 0 : NGHTTP2_ERR_START_STREAM_NOT_ALLOWED;
diff --git a/modules/http2/h2_workers.c b/modules/http2/h2_workers.c
index 28bb428..ae250b0 100644
--- a/modules/http2/h2_workers.c
+++ b/modules/http2/h2_workers.c
@@ -479,8 +479,6 @@ apr_status_t h2_workers_unregister(h2_workers *workers, struct h2_mplx *m)
 void h2_workers_graceful_shutdown(h2_workers *workers)
 {
     workers->shutdown = 1;
-    workers->min_workers = 1;
     workers->max_idle_duration = apr_time_from_sec(1);
-    h2_fifo_term(workers->mplxs);
     wake_non_essential_workers(workers);
 }
